src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java

```diff
@@ -207,10 +213,36 @@ public static long getKey0(int refIdx, int alignmentStart0) {
 				}
 			}
 
+			if (!overlaps(r, intervals)) {
+				continue;
+			}
+
 			key.set(getKey(r));
 			record.set(r);
 			return true;
 		}
 		return false;
 	}
+
+	private static boolean overlaps(SAMRecord r, List<Locatable> intervals) {
+		if (intervals == null) {
```

Comment:
Maybe switch the condition to <pre>intervals == null || (r.getReadUnmappedFlag() && r.getAlignmentStart() == SAMRecord.NO_ALIGNMENT_START)</pre>? Then the check below could be slightly simplified but the benefit would be to completely skip interval checking for unmapped reads.